README.md
# NAME

POE::Component::MessageQueue - A Perl message queue based on POE that uses STOMP as its communication protocol.

# USAGE

If you are only interested in running with the recommended storage backend and
some predetermined defaults, you can use the included command line script:

    POE::Component::MessageQueue version 0.2.12
    Copyright 2007-2011 David Snopek (http://www.hackyourlife.org)
    Copyright 2007, 2008 Paul Driver <frodwith@gmail.com>
    Copyright 2007 Daisuke Maki <daisuke@endeworks.jp>

    mq.pl [--port|-p <num>]               [--hostname|-h <host>]
          [--front-store <str>]           [--front-max <size>] 
          [--granularity <seconds>]       [--nouuids]
          [--timeout|-i <seconds>]        [--throttle|-T <count>]
		[--dbi-dsn <str>]               [--mq-id <str>]
		[--dbi-username <str>]          [--dbi-password <str>]
          [--pump-freq|-Q <seconds>]
          [--data-dir <path_to_dir>]      [--log-conf <path_to_file>]
          [--stats-interval|-i <seconds>] [--stats]
          [--pidfile|-p <path_to_file>]   [--background|-b]
          [--crash-cmd <path_to_script>]
          [--debug-shell] [--version|-v]  [--help|-h]

    SERVER OPTIONS:
      --port     -p <num>     The port number to listen on (Default: 61613)
      --hostname -h <host>    The hostname of the interface to listen on 
                              (Default: localhost)

    STORAGE OPTIONS:
      --storage <str>         Specify which overall storage engine to use.  This
                              affects what other options are value.  (can be
                              default or dbi)
      --front-store -f <str>  Specify which in-memory storage engine to use for
                              the front-store (can be memory or bigmemory).
      --front-max <size>      How much message body the front-store should cache.
                              This size is specified in "human-readable" format
                              as per the -h option of ls, du, etc. (ex. 2.5M)
      --timeout -i <secs>     The number of seconds to keep messages in the 
                              front-store (Default: 4)
      --pump-freq -Q <secs>   How often (in seconds) to automatically pump each
                              queue.  Set to zero to disable this timer entirely
                              (Default: 0)
      --granularity <secs>    How often (in seconds) Complex should check for
                              messages that have passed the timeout.  
      --[no]uuids             Use (or do not use) UUIDs instead of incrementing
                              integers for message IDs.  (Default: uuids)
      --throttle -T <count>   The number of messages that can be stored at once 
                              before throttling (Default: 2)
      --data-dir <path>       The path to the directory to store data 
                              (Default: /var/lib/perl_mq)
      --log-conf <path>       The path to the log configuration file 
                              (Default: /etc/perl_mq/log.conf)

    --dbi-dsn <str>         The database DSN when using --storage dbi
    --dbi-username <str>    The database username when using --storage dbi
    --dbi-password <str>    The database password when using --storage dbi
    --mq-id <str>           A string uniquely identifying this MQ when more
                            than one MQ use the DBI database for storage

    STATISTICS OPTIONS:
      --stats                 If specified the, statistics information will be 
                              written to $DATA_DIR/stats.yml
      --stats-interval <secs> Specifies the number of seconds to wait before 
                              dumping statistics (Default: 10)

    DAEMON OPTIONS:
      --background -b         If specified the script will daemonize and run in the
                              background
      --pidfile    -p <path>  The path to a file to store the PID of the process

    --crash-cmd  <path>     The path to a script to call when crashing.
                            A stacktrace will be printed to the script's STDIN.
                            (ex. 'mail root@localhost')

    OTHER OPTIONS:
      --debug-shell           Run with POE::Component::DebugShell
      --version    -v         Show the current version.
      --help       -h         Show this usage message

# SYNOPSIS

## Subscriber

    use Net::Stomp;
    

    my $stomp = Net::Stomp->new({
      hostname => 'localhost',
      port     => 61613
    });
    

    # Currently, PoCo::MQ doesn't do any authentication, so you can put
    # whatever you want as the login and passcode.
    $stomp->connect({ login => $USERNAME, passcode => $PASSWORD });
    

    $stomp->subscribe({
      destination => '/queue/my_queue.sub_queue',
      ack         => 'client'
    });
    

    while (1)
    {
      my $frame = $stomp->receive_frame;
      print $frame->body . "\n";
      $stomp->ack({ frame => $frame });
    }
    

    $stomp->disconnect();

## Producer

    use Net::Stomp;
    

    my $stomp = Net::Stomp->new({
      hostname => 'localhost',
      port     => 61613
    });
    

    # Currently, PoCo::MQ doesn't do any authentication, so you can put
    # whatever you want as the login and passcode.
    $stomp->connect({ login => $USERNAME, passcode => $PASSWORD });
    

    $stomp->send({
      destination => '/queue/my_queue.sub_queue',
      body        => 'I am a message',
      persistent  => 'true',
    });
    

    $stomp->disconnect();

## Server

If you want to use a different arrangement of storage engines or to embed PoCo::MQ
inside another application, the following synopsis may be useful to you:

    use POE;
    use POE::Component::Logger;
    use POE::Component::MessageQueue;
    use POE::Component::MessageQueue::Storage::Default;
    use Socket; # For AF_INET
    use strict;

    my $DATA_DIR = '/tmp/perl_mq';

    # we create a logger, because a production message queue would
    # really need one.
    POE::Component::Logger->spawn(
      ConfigFile => 'log.conf',
      Alias      => 'mq_logger'
    );

    POE::Component::MessageQueue->new({
      port     => 61613,            # Optional.
      address  => '127.0.0.1',      # Optional.
      hostname => 'localhost',      # Optional.
      domain   => AF_INET,          # Optional.
      

    logger_alias => 'mq_logger',  # Optional.

      # Required!!
      storage => POE::Component::MessageQueue::Storage::Default->new({
        data_dir     => $DATA_DIR,
        timeout      => 2,
        throttle_max => 2
      })
    });

    POE::Kernel->run();
    exit;

# DESCRIPTION

This module implements a message queue \[1\] on top of [POE](http://search.cpan.org/perldoc?POE) that communicates
via the STOMP protocol \[2\].

There exist a few good Open Source message queues, most notably ActiveMQ \[3\] which
is written in Java.  It provides more features and flexibility than this one (while
still implementing the STOMP protocol), however, it was (at the time I last used it)
very unstable.  With every version there was a different mix of memory leaks, persistence
problems, STOMP bugs, and file descriptor leaks.  Due to its complexity I was
unable to be very helpful in fixing any of these problems, so I wrote this module!

This component distinguishes itself in a number of ways:

- No OS threads, its asynchronous.  (Thanks to [POE](http://search.cpan.org/perldoc?POE)!)
- Persistence was a high priority.
- A strong effort is put to low memory and high performance.
- Message storage can be provided by a number of different backends.
- Features to support high-availability and fail-over.  (See the ["\#HIGH AVAILABILITY"](#\#HIGH AVAILABILITY) section below)

## Special STOMP headers

You can see the main STOMP documentation here: [http://stomp.codehaus.org/Protocol](http://stomp.codehaus.org/Protocol)

PoCo::MQ implements a number of non-standard STOMP headers:

- __persistent__

    Set to the string "true" to request that a message be persisted.  Not setting this header
    or setting it to any other value, means that a message is non-persistent.

    Many storage engines ignore the "persistent" header, either persisting all messages or 
    no messages, so be sure to check the documentation for your storage engine.

    Using the Complex or Default storage engines, persistent messages will always be sent
    to the back store and non-persistent messages will be discarded eventually.

- __expire-after__

    For non-persistent messages, you can set this header to the number of seconds this
    message must be kept before being discarded.  This is ignored for persistent messages.

    Many storage engines ignore the "expire-after" header, so be sure to check the
    documentation for your storage engine.

    Using the Complex or Default storage engines, this header will be honored.  If it isn't
    specified, non-persistent messages are discarded when pushed out of the front store.

- __deliver-after__

    For both persistent or non-persistent messages, you can set this header to the number of
    seconds this message should be held before being delivered.  In other words, this allows
    you to delay delivery of a message for an arbitrary number of seconds.

    All the storage engines in the standard distribution support this header.  __But it will not
    work without a pump frequency enabled!__  If using mq.pl, enable with --pump-freq or if creating
    a [POE::Component::MessageQueue](http://search.cpan.org/perldoc?POE::Component::MessageQueue) object directly, pass pump\_frequency as an argument to new().

## Queues and Topics

In PoCo::MQ there are two types of _destinations_: __queues__ and __topics__

- __queue__

    Each message is only delivered to a single subscriber (not counting
    messages that were delivered but not ACK'd).  If there are multiple
    subscribers on a single queue, the messages will be divided amoung them,
    roughly equally.

- __topic__

    Each message is delivered to every subscriber.  Topics don't support any kind
    of persistence, so to get a message, a subscriber _must_ be connected at the
    time it was sent.

All destination names start with either "/queue/" or "/topic/" to distinguish
between queues and topics.

## Tips and Tricks

- __Logging!  Use it.__

    PoCo::MQ uses [POE::Component::Logger](http://search.cpan.org/perldoc?POE::Component::Logger) for logging which is based on
    [Log::Dispatch](http://search.cpan.org/perldoc?Log::Dispatch).  By default __mq.pl__ looks for a log file at:
    "/etc/perl\_mq/log.conf".  Or you can specify an alternate location with the
    _\--log-conf_ command line argument.  

- __Using the login/passcode to track clients in the log.__

    Currently the login and passcode aren't used by PoCo::MQ for auth, but they
    _are_ written to the log file.  In the log file clients are only identified
    by the client id.  But if you put information identifying the client in the
    login/passcode you can connect that to a client id by finding it in the log.

# STORAGE

When creating an instance of this component you must pass in a storage object
so that the message queue knows how to store its messages.  There are some storage
backends provided with this distribution.  See their individual documentation for 
usage information.  Here is a quick break down:

- [POE::Component::MessageQueue::Storage::Memory](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Memory) -- The simplest storage engine.  It keeps messages in memory and provides absolutely no presistence.
- [POE::Component::MessageQueue::Storage::BigMemory](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::BigMemory) -- An alternative memory storage engine that is optimized for large numbers of messages.
- [POE::Component::MessageQueue::Storage::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::DBI) -- Uses Perl [DBI](http://search.cpan.org/perldoc?DBI) to store messages.  Depending on your database configuration, using directly may not be recommended because the message bodies are stored in the database.  Wrapping with [POE::Component::MessageQueue::Storage::FileSystem](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::FileSystem) allows you to store the message bodies on disk.  All messages are stored persistently.  (Underneath this is really just [POE::Component::MessageQueue::Storage::Generic](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic) and [POE::Component::MessageQueue::Storage::Generic::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic::DBI))
- [POE::Component::MessageQueue::Storage::FileSystem](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::FileSystem) -- Wraps around another storage engine to store the message bodies on the filesystem.  This can be used in conjunction with the DBI storage engine so that message properties are stored in DBI, but the message bodies are stored on disk.  All messages are stored persistently regardless of whether a message has set the persistent header or not.
- [POE::Component::MessageQueue::Storage::Generic](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic) -- Uses [POE::Component::Generic](http://search.cpan.org/perldoc?POE::Component::Generic) to wrap storage modules that aren't asynchronous.  Using this module is the easiest way to write custom storage engines.
- [POE::Component::MessageQueue::Storage::Generic::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic::DBI) -- A synchronous [DBI](http://search.cpan.org/perldoc?DBI)\-based storage engine that can be used inside of Generic.  This provides the basis for the [POE::Component::MessageQueue::Storage::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::DBI) module.
- [POE::Component::MessageQueue::Storage::Throttled](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Throttled) -- Wraps around another engine to limit the number of messages sent to be stored at once.  Use of this module is __highly__ recommended!  If the storage engine is unable to store the messages fast enough (ie. with slow disk IO) it can get really backed up and stall messages coming out of the queue, allowing execessive producers to basically monopolize the server, preventing any messages from getting distributed to subscribers.  Also, it will significantly cuts down the number of open FDs when used with [POE::Component::MessageQueue::Storage::FileSystem](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::FileSystem).  Internally it makes use of [POE::Component::MessageQueue::Storage::BigMemory](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::BigMemory) to store the throttled messages.
- [POE::Component::MessageQueue::Storage::Complex](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Complex) -- A configurable storage engine that keeps a front-store (something fast) and a back-store (something persistent), allowing you to specify a timeout and an action to be taken when messages in the front-store expire, by default, moving them into the back-store.  This optimization allows for the possibility of messages being handled before ever having to be persisted.  Complex is capable to correctly handle the persistent and expire-after headers.
- [POE::Component::MessageQueue::Storage::Default](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Default) -- A combination of the Complex, BigMemory, FileSystem, DBI and Throttled modules above.  It will keep messages in BigMemory and move them into FileSystem after a given number of seconds, throttling messages passed into DBI.  The DBI backend is configured to use SQLite.  It is capable to correctly handle the persistent and expire-after headers.  This is the recommended storage engine and should provide the best performance in the most common case (ie. when both providers and consumers are connected to the queue at the same time).

# CONSTRUCTOR PARAMETERS

- storage => SCALAR

    The only required parameter.  Sets the object that the message queue should use for
    message storage.  This must be an object that follows the interface of
    [POE::Component::MessageQueue::Storage](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage) but doesn't necessarily need to be a child
    of that class.

- alias => SCALAR

    The session alias to use.

- port => SCALAR

    The optional port to listen on.  If none is given, we use 61613 by default.

- address => SCALAR

    The option interface address to bind to.  It defaults to INADDR\_ANY or INADDR6\_ANY
    when using IPv4 or IPv6, respectively.

- hostname => SCALAR

    The optional name of the interface to bind to.  This will be converted to the IP and
    used as if you set _address_ instead.  If you set both _hostname_ and _address_,
    _address_ will override this value.

- domain => SCALAR

    Optionally specifies the domain within which communication will take place.  Defaults
    to AF\_INET.

- logger\_alias => SCALAR

    Optionally set the alias of the POE::Component::Logger object that you want the message
    queue to log to.  If no value is given, log information is simply printed to STDERR.

- message\_class => SCALAR

    Optionally set the package name to use for the Message object.  This should be a child
    class of POE::Component::MessageQueue::Message or atleast follow the same interface.

    This allows you to add new message headers which the MQ can recognize.

- pump\_frequency => SCALAR

    Optionally set how often (in seconds) to automatically pump each queue.  If zero or
    no value is given, then this timer is disabled entirely.

    When disabled, each queue is only pumped when its contents change, meaning 
    when a message is added or removed from the queue.  Normally, this is enough.  However,
    if your storage engine holds back messages for any reason (ie. to delay their 
    delivery) it will be necessary to enable this, so that the held back messages will
    ultimately be delivered.

    _You must enable this for the message queue to honor the deliver-after header!_

- observers => ARRAYREF

    Optionally pass in a number of objects that will receive information about events inside
    of the message queue.

    Currently, only one observer is provided with the PoCo::MQ distribution:
    [POE::Component::MessageQueue::Statistics](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Statistics).  Please see its documentation for more information.

# HIGH AVAILABILITY

From version 0.2.10, PoCo::MQ supports a features to enable high availability.

- __Clustering__

    You can now run multiple MQs which share the same back-store, behind a reverse-proxy load-balancer with
    automatic fail-over, if one of the MQs goes down.

    See the the clustering documentation for more information:

    [POE::Component::MessageQueue::Manual::Clustering](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Manual::Clustering)

- __DBI fail-over__

    The DBI storage engine can be configured with a list of database servers.  If one of them is not available
    or goes down, it will fail-over to the next one.

    If you set up several database servers with master-to-master replication, this will allow the MQ to seemlessly
    handle failure of one of the databases.

    See the DBI storage engine documentation for more information:

    [POE::Component::MessageQueue::Storage::Generic::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic::DBI)

# REFERENCES

- \[1\]

    [http://en.wikipedia.org/wiki/Message\_Queue](http://en.wikipedia.org/wiki/Message\_Queue) -- General information about message queues

- \[2\]

    [http://stomp.codehaus.org/Protocol](http://stomp.codehaus.org/Protocol) -- The informal "spec" for the STOMP protocol

- \[3\]

    [http://www.activemq.org/](http://www.activemq.org/) -- ActiveMQ is a popular Java-based message queue

# UPGRADING FROM OLDER VERSIONS

If you used any of the following storage engines with PoCo::MQ 0.2.9 or older:

- [POE::Component::MessageQueue::Storage::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::DBI)

The database format has changed!

__Note:__ When using [POE::Component::MessageQueue::Storage::Default](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Default) (meaning mq.pl
\--storage default) the database will be automatically updated in place, so you don't
need to worry about this.

Included in the distribution, is a schema/ directory with a few SQL scripts for 
upgrading:

- upgrade-0.1.7.sql -- Apply if you are upgrading from version 0.1.6 or older.
- upgrade-0.1.8.sql -- Apply if your are upgrading from version 0.1.7 or after applying
the above upgrade script.  This one has a SQLite specific version: upgrade-0.1.8-sqlite.sql).
- upgrade-0.2.3.sql -- Apply if you are upgrading from version 0.2.2 or older (after
applying the above upgrade scripts).
- upgrade-0.2.9-mysql.sql -- Doesn't apply to SQLite users!  Apply if you are upgrading from version
0.2.8 or older (after applying the above upgrade scripts).
- upgrade-0.2.10-mysql.sql -- Doesn't apply to SQLite users!  Apply if you are upgrading from version
0.2.9 or older (after applying the above upgrade scripts).

# CONTACT

Please check out the Google Group at:

[http://groups.google.com/group/pocomq](http://groups.google.com/group/pocomq)

Or just send an e-mail to: pocomq@googlegroups.com

# DEVELOPMENT

If you find any bugs, have feature requests, or wish to contribute, please
contact us at our Google Group mentioned above.  We'll do our best to help you
out!

Development is coordinated via Bazaar (See [http://bazaar-vcs.org](http://bazaar-vcs.org)).  The main
Bazaar branch can be found here:

[http://code.hackyourlife.org/bzr/dsnopek/perl\_mq/devel.mainline](http://code.hackyourlife.org/bzr/dsnopek/perl\_mq/devel.mainline)

We prefer that contributions come in the form of a published Bazaar branch with the
changes.  This helps facilitate the back-and-forth in the review process to get
any new code merged into the main branch.

There is also an official git mirror hosted on GitHub here:

[https://github.com/dsnopek/POE--Component--MessageQueue](https://github.com/dsnopek/POE--Component--MessageQueue)

We will also accept contributions via git and GitHub pull requests!

# FUTURE

The goal of this module is not to support every possible feature but rather to
be small, simple, efficient and robust.  For the most part expect incremental
changes to address those areas.

Beyond that we have a TODO list (shown below) called __"The Long Road To
1.0"__.  This is a list of things we feel we need to have inorder to call the
product complete.  That includes management and monitoring tools for sysadmins
as well as documentation for developers.

- __Full support for STOMP__: Includes making sure we are robust to clients
participating badly in the protocol.
- __Authentication and authorization__: This should be highly pluggable, but
basically (as far as authorization goes) each user can get read/write/admin
perms for a queue which are inherited by default to sub-queues (as separated
by the dot character).
- __Monitoring/management tools__:  It should be possible for an admin to monitor the
overall state of the queue, ie: (1) how many messages for what queues are in
the front-store, throttled, back-store, etc, (2) information on connected
clients, (3) data/message thorough put, (4) daily/weekly/monthly trends, (X)
etc..  They should also be able to "peek" at any message at any point as well
as delete messages or whole queues.
The rough plan is to use special STOMP frames and "magic" queues/topics to
access special information or perform admin tasks.  Command line scripts for
simple things would be included in the main distribution and a full-featured
web-interface would be provided as a separate module.
- __Log rotation__: At minimum, documentation on how to set it up.
- __Docs on "using" the MQ__: A full tutorial from start to finish, advice on
writing good consumers/producers and solid docs on authoring custom storage
engines.

# APPLICATIONS USING PoCo::MQ

- [http://chessvegas.com](http://chessvegas.com)

    Chess gaming site ChessVegas.

# SEE ALSO

_External modules:_

[POE](http://search.cpan.org/perldoc?POE),
[POE::Component::Server::Stomp](http://search.cpan.org/perldoc?POE::Component::Server::Stomp),
[POE::Component::Client::Stomp](http://search.cpan.org/perldoc?POE::Component::Client::Stomp),
[Net::Stomp](http://search.cpan.org/perldoc?Net::Stomp),
[POE::Filter::Stomp](http://search.cpan.org/perldoc?POE::Filter::Stomp),
[POE::Component::Logger](http://search.cpan.org/perldoc?POE::Component::Logger),
[DBD::SQLite](http://search.cpan.org/perldoc?DBD::SQLite),
[POE::Component::Generic](http://search.cpan.org/perldoc?POE::Component::Generic)

_Storage modules:_

[POE::Component::MessageQueue::Storage](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage),
[POE::Component::MessageQueue::Storage::Memory](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Memory),
[POE::Component::MessageQueue::Storage::BigMemory](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::BigMemory),
[POE::Component::MessageQueue::Storage::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::DBI),
[POE::Component::MessageQueue::Storage::FileSystem](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::FileSystem),
[POE::Component::MessageQueue::Storage::Generic](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic),
[POE::Component::MessageQueue::Storage::Generic::DBI](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Generic::DBI),
[POE::Component::MessageQueue::Storage::Double](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Double),
[POE::Component::MessageQueue::Storage::Throttled](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Throttled),
[POE::Component::MessageQueue::Storage::Complex](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Complex),
[POE::Component::MessageQueue::Storage::Default](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Storage::Default)

_Statistics modules:_

[POE::Component::MessageQueue::Statistics](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Statistics),
[POE::Component::MessageQueue::Statistics::Publish](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Statistics::Publish),
[POE::Component::MessageQueue::Statistics::Publish::YAML](http://search.cpan.org/perldoc?POE::Component::MessageQueue::Statistics::Publish::YAML)

_ID generator modules:_

[POE::Component::MessageQueue::IDGenerator](http://search.cpan.org/perldoc?POE::Component::MessageQueue::IDGenerator),
[POE::Component::MessageQueue::IDGenerator::SimpleInt](http://search.cpan.org/perldoc?POE::Component::MessageQueue::IDGenerator::SimpleInt),
[POE::Component::MessageQueue::IDGenerator::UUID](http://search.cpan.org/perldoc?POE::Component::MessageQueue::IDGenerator::UUID)

# BUGS

We are serious about squashing bugs!  Currently, there are no known bugs, but
some probably do exist.  If you find any, please let us know at the Google group.

That said, we are using this in production in a commercial application for
thousands of large messages daily and we experience very few issues.

# AUTHORS

Copyright 2007-2011 David Snopek ([http://www.hackyourlife.org](http://www.hackyourlife.org))

Copyright 2007, 2008 Paul Driver <frodwith@gmail.com>

Copyright 2007 Daisuke Maki <daisuke@endeworks.jp>

# LICENSE

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGES

== 0.2.14 -- 06/30/2013 ==

 * Migrated to Dist::Zilla

== 0.2.13 -- 06/03/2012 ==

 * Retry only max_retry times when a db operation fails.
 * Fixed deprecated usage of qw() as parentheses.

== 0.2.12 -- 07/28/2011 ==

 * Added missing test which caused last release to fail to install (Thanks to Naveed
   Massjouni!)

== 0.2.11 -- 07/26/2011 ==

 * Unit tests and documentation for the clustering support added in the last release.
 * Updated to work with the latest Moose (2.0201)

== 0.2.10 -- 07/20/2010 ==

 * Added Paul Driver's Storage::Remote (a remote front-store for a clustered MQ)
 * Multiple Storage::DBI engines in seperate MQ's can now share the same DB if they are
   given unique mq_id's.
 * Storage::DBI can now take a list of DB servers and will attempt to failover to another
   one when the current one goes down.
 * mq.pl can now be used to run an MQ based on Storage::DBI.
 * DB schemas changed.

== 0.2.9 -- 06/05/2010 ==

 * A fix to bug #49666 which affected the order of messages (Thanks Naveed Massjouni!)
 * Depends on POE::Filter::Stomp 0.04 or later (fixes issues with Windows-style newlines)
 * Fixed a regression in Storage::DBI->new( ... ) - Can now pass a HASH ref
 * Fixed schemas for MySQL

== 0.2.8 -- 01/26/2010 ==

 * A better fix for new POE which doesn't crash the tests on some users systems.

== 0.2.7 -- 01/23/2010 ==

 * Fixes to get PoCo::MQ working with latest Moose and POE (Thanks to SQUEEK!)

== 0.2.6 -- 02/04/2009 ==

 * Possible fix for a bug where PoCo::MQ crashes because a message is being ACK'd by a client
   that isn't the one the message was originally sent to and the original client is now
   disconnected.
 * Fixed a bug (since 0.2.0) where unsubscribing from a queue wouldn't correctly disown a
   subscribers owned messages (Thanks Jason McManus!)
 * Small fixes and additions to documentation and unit tests.

== 0.2.5 -- 12/18/2008 ==

 * Fixed dependancies in Makefile.PL per rt# 41786 (Thanks Sebastien Aperghis-Tramoni!)

== 0.2.4 -- 10/27/2008 ==

 * Fix for bug where delayed messages could be delivered more than once to different subscribers.
 * Changed 'receipt' header on RECEIPT frames to 'receipt-id' to match the STOMP spec.

== 0.2.3 -- 10/9/2008 ==

 * Added the 'message_class' argument to allow extending the message object.
 * Added the 'pump_frequency' argument to enable an automatic pump timer (also --pump-freq to mq.pl).
 * Added support for the 'deliver-after' header to delay message delivery (db schema changes!).

== 0.2.2 -- 6/6/2008 ==

 * Fix for a regression in 0.2.x with regard to ack => 'auto'.

== 0.2.1 -- 4/28/2008 ==

 * Unbroke the interface to Storage::Generic.
 * Added missing dependancy to Data::UUID.

== 0.2.0 -- 4/24/2008 ==

 * Massive refactor:
   - Built on Moose.
   - Any storage engine can be a front-store.
   - Callback passing (versus old static method).
 * Storage::Complex:
   - Treat the front store like a cache.
   - Separate 'granularity' parameter.
 * Storage::Throttled:
   - Can claim and retrieve messages for improved performance.
 * Comprehensive unit-test suite.

== 0.1.8 -- 2/5/2008 ==

 * Configurable message ID generators (UUIDs are now the default).
 * Abstracted Storage::Complex to take a configurable front-store and expire function.
 * An alterative memory-based storage engine (Storage::BigMemory) which operates in
   O(1) time with any number of messages.
 * Support for non-persistent topics (via topic://).
 * Use POE::Filter::Stomp for handling STOMP frame parsing.

== 0.1.7 -- 12/19/2007 ==

 * Added support for statistics gathering.
 * Attempt to shutdown gracefully after receiving the TERM, INT or HUP signals.
 * Fixed a bug where messages could get owned by a subscriber forever.

== 0.1.6 -- 9/5/2007 ==

 * Added a pump_deferred() function to send a message to pump a queue
 * Fixed a bug with ack_type => 'auto'
 * mq.pl: Output a sane error message when unable to write the crashed.log

== 0.1.5 -- 8/27/2007 ==

 * Added missing dependancies to Makefile.PL

== 0.1.4 -- 6/25/2007 ==

 * Added crash.log and daemonizing support to mq.pl script.
 * Added documentation for mq.pl script.
 * Various little documentation fixes.



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A STOMP based message queue server",
  "AUTHOR" => "David Snopek (http://www.hackyourlife.org), Paul Driver <frodwith\@gmail.com>, Daisuke Maki <daisuke\@endeworks.jp>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "POE-Component-MessageQueue",
  "EXE_FILES" => [
    "bin/mq.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "POE::Component::MessageQueue",
  "PREREQ_PM" => {
    "Best" => 0,
    "Carp" => 0,
    "DBD::SQLite" => "1.13",
    "DBI" => 0,
    "Data::Dumper" => 0,
    "Data::UUID" => 0,
    "Event::Notify" => 0,
    "Exception::Class::DBI" => 0,
    "Exception::Class::TryCatch" => 0,
    "File::Temp" => 0,
    "Heap" => "0.80",
    "Heap::Fibonacci" => "0.80",
    "IO::File" => 0,
    "IO::String" => 0,
    "Moose" => "2.0201",
    "MooseX::MultiInitArg" => 0,
    "Net::Stomp" => 0,
    "POE" => "0.38",
    "POE::Component::Generic" => "0.1001",
    "POE::Component::Logger" => 0,
    "POE::Component::Server::TCP" => 0,
    "POE::Filter::Stomp" => "0.04",
    "POE::Wheel::ReadWrite" => 0,
    "Test::Exception" => 0,
    "Test::MockObject" => 0,
    "Test::More" => 0,
    "YAML" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.3000",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



